# Исследование надежности заемщиков

Цель: На основе статистики о платёжеспособности клиентов исследовать, влияет ли семейное положение, количество детей клиента, ежемесячный доход и цели получения кредита на факт возврата кредита в срок

**ОПИСАНИЕ ДАННЫХ**

- children — количество детей в семье
- days_employed — общий трудовой стаж в днях
- dob_years — возраст клиента в годах
- education — уровень образования клиента
- education_id — идентификатор уровня образования
- family_status — семейное положение
- family_status_id — идентификатор семейного положения
- gender — пол клиента
- income_type — тип занятости
- debt — имел ли задолженность по возврату кредитов
- total_income — ежемесячный доход
- purpose — цель получения кредита


## Импорт библиотек и изучение общей информации о данных

In [1]:
import pandas as pd

In [2]:
try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

In [3]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

Посмотрим количество пропущенных значений для каждого столбца

In [5]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

В двух столбцах есть пропущенные значения. В столбце с данными о доходах `total_income` заполним пропуски медианным значением по каждому типу из столбца `income_type`, так как на сумму дохода сильнее всего влияет тип занятости.

In [6]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

Заполним пропуски в столбце days_employed медианными значениями по каждого типа занятости income_type

In [12]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

Проверим, что все пропуски заполнены

In [13]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Обработка аномальных значений

В столбце `days_employed`, количество дней трудового стажа, есть аномальные отрицательные значения. Заменим их на положительные.

In [7]:
data['days_employed'] = data['days_employed'].abs()

Для каждого типа занятости выведем медианное значение трудового стажа days_employed в днях

In [8]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

В двух типах (безработные и пенсионеры) присутствуют аномально большие значения. Оставим их как есть, так как этот столбец не понадобится для исследования.

Выведем перечень уникальных значений столбца `children`

In [9]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

Удалим строки с аномальными значениями.

In [10]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

Проверим результат

In [11]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Изменение типов данных

Заменим вещественный тип данных в столбце total_income на целочисленный

In [14]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

В столбце `education` есть одни и те же значения, записанные по-разному. Приведем данные к нижнему регистру.

In [15]:
data['education'] = data['education'].str.lower()

Проверим данные на количество строк-дубликатов. Если такие строки присутствуют, удалим

In [16]:
data.duplicated().sum()

71

In [19]:
data = data.drop_duplicates()

### Категоризация данных

На основании диапазонов, указанных ниже, создадим в датафрейме столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


In [20]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [21]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

Посмотрим на перечень уникальных целей взятия кредита из столбца `purpose`

In [22]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

Создадим функцию, которая на основании данных из столбца purpose сформирует новый столбец purpose_category, в который войдут следующие категории:

- 'операции с автомобилем',
- 'операции с недвижимостью',
- 'проведение свадьбы',
- 'получение образования'.

In [23]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [24]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

## Исследование  данных и ответы на поставленные вопросы

### Зависимость между количеством детей и возвратом кредита в срок

In [25]:
debt_children = pd.DataFrame()

debt_children['count_children'] = data.groupby('children')['debt'].count()

debt_children['sum_children'] = data.groupby('children')['debt'].sum() 

debt_children['result_children'] = data.groupby('children')['debt'].mean()  

debt_children.sort_values('result_children', ascending = False)


count_children  sum_children  result_children
children                                               
4                     41             4         0.097561
2                   2052           194         0.094542
1                   4808           444         0.092346
3                    330            27         0.081818
0                  14091          1063         0.075438
5                      9             0         0.000000

**Вывод:** Полученные данные неоднозначны, возможно, нужна бОльшая выборка. Только 9 человек имеют пятерых детей, этого не достаточно, чтобы делать выводы по данной категории. В имеющейся выборке эта категория не оформляла кредитов. С увеличением количества детей в семье люди реже обращаются за кредитами. Наибольшее количество кредитов берут и имеют меньше просрочек бездетные люди. Просроченные выплаты примерно одинаковы у семей с 1,2 и 4-мя деьтми, хотя данных о семьях с 4-мя детьми также недостаточно. Из семей с детьми наиболее часто осуществляют выплаты в срок семьи с 3-мя детьми.

### Зависимость между семейным положением и возвратом кредита в срок

In [26]:
debt_family_status = pd.DataFrame()

debt_family_status['sum_family_status'] = data.groupby('family_status')['debt'].sum() 

debt_family_status['count_family_status'] = data.groupby('family_status')['debt'].count() 

debt_family_status['result_family_status'] = data.groupby('family_status')['debt'].mean() 

debt_family_status.sort_values('result_family_status', ascending = False)


sum_family_status  count_family_status  \
family_status                                                   
Не женат / не замужем                273                 2796   
гражданский брак                     385                 4134   
женат / замужем                      927                12261   
в разводе                             84                 1189   
вдовец / вдова                        63                  951   

                       result_family_status  
family_status                                
Не женат / не замужем              0.097639  
гражданский брак                   0.093130  
женат / замужем                    0.075606  
в разводе                          0.070648  
вдовец / вдова                     0.066246

**Вывод:** Наблюдается зависимость платежеспособности от семейного положения. Люди, проживающие в гражданском браке и не бывавшие в браке, имеют бОльшее количество задежек выплат. Холостые/не замужние часто берут кредиты. Наименьшее колчичество просроченных платежей среди овдовевших, однако в данной категории очень малое количество взятых кредитов. Также редко берут кредиты люди в разводе. Люди, состоящие в браке, наиболее часто берут кредиты и чаще осуществляют выплаты кредитов в срок.

### Зависимость между уровнем дохода и возвратом кредита в срок

In [27]:
debt_income = pd.DataFrame()

debt_income['count_total_income'] = data.groupby('total_income_category')['debt'].count()

debt_income['sum_total_income'] = data.groupby('total_income_category')['debt'].sum()

debt_income['result_total_income'] = debt_income['sum_total_income'] / debt_income['count_total_income']

debt_income.sort_values('result_total_income', ascending = False)
                    

count_total_income  sum_total_income  \
total_income_category                                         
E                                      22                 2   
C                                   15921              1353   
A                                      25                 2   
B                                    5014               354   
D                                     349                21   

                       result_total_income  
total_income_category                       
E                                 0.090909  
C                                 0.084982  
A                                 0.080000  
B                                 0.070602  
D                                 0.060172

**Вывод:** Зависимость между уровнем дохода и вовзратом кредитов в срок очень незначительна. Значительно бОльшее количество кредитов оформляют люди с доходом от 50 до 200 000 руб. Также оформляют кредиты люди с доходом от 200 000 до 1 млн и реже - от 50 до 200 000. Люди с очень высоким (более 1 млн) и очень низким (до 50 000) доходом оформляют кредиты крайне мало. Для данных категорий нужна ббОльшая выборка.

### Как разные цели кредита влияют на его возврат в срок

In [28]:
debt_purpose_category = pd.DataFrame()

debt_purpose_category['sum_purpose_category'] = data.groupby('purpose_category')['debt'].sum()

debt_purpose_category['count_purpose_category'] = data.groupby('purpose_category')['debt'].count()

debt_purpose_category['result_purpose_category'] = debt_purpose_category['sum_purpose_category']
/ debt_purpose_category['count_purpose_category'] 

debt_purpose_category.sort_values('result_purpose_category', ascending = False)

sum_purpose_category  count_purpose_category  \
purpose_category                                                         
операции с недвижимостью                   780                   10751   
операции с автомобилем                     400                    4279   
получение образования                      369                    3988   
проведение свадьбы                         183                    2313   

                          result_purpose_category  
purpose_category                                   
операции с недвижимостью                      780  
операции с автомобилем                        400  
получение образования                         369  
проведение свадьбы                            183

**Вывод:** Значительно больше количество кредитов, примерно в два раза, оформляется на операции с недвижимостью. Меньше всего оформлено кредитов на проведение свадьбы. Чаще всего платят в срок с целью кредита недвижимость и на свадьбу. Большее колчиество просроченных выплат у тех, кому нужен кредит на операции с автомобилем и образование.

## Общий вывод

**Выводы:** 

Проведено исследование влияния семейного положения, количества детей, ежемесячного дохода и цели получения кредита на надежность заемщиков и погашение кредитных обязательств в срок. В качестве вводных данных использована выборка данных кредитного отдела банка по статистике о платежеспособности клиентов.
В данных обнаружены пропуски, причинами их повления могут быть технологические проблемы или человеческий фактор (опечатки, нежелание оставлять информацию и т.д.). Также пропуски могут быть связаны с неофициальным трудоустройством респондентов.

В результате исследования удалось выявить следующие закономерности:
1) Влияние количества детей на сроки погашения кредита. 
Обнаружены следующие закономерности: наибольшее количество кредитов берут и выплачивают в срок бездетные люди (7,5% просрочек при 1063 оформленных кредитов). На данную категорию приходится почти в три раза больше оформленных кредитов по сравнению с другими категориями.
С увеличением количества детей в семье люди реже обращаются за кредитами.
Можно говорить , что количество просроченных выплат примерно одинаково у людей с 1 и 2-ми детьми (9,2% и 9,5%). Семьи с 3-мя детьми имеют меньший процент просрочек (8,2%) однако в выборке недостаточно данных по этой категории. Также необходимо больше данных по семьям с 4-мя и 5-тью детьми. 
В рамках имеющихся данных сделать однозначные выводы нельзя, необходима бОльшая информация по семьям с 3-мя, 4-мя и 5-тью детьми.

2) Влияние семейного статуса на сроки погашения кредита.
По сравнению с прочими проанализированными показателями на сроки выплат в значительной степени влияет семейный статус. Люди состоящие в браке, наиболее часто оформляют кредиты и чаще осуществляют выплаты вовремя (7,6% просроченных выплат).  
Люди, проживающие в гражданском браке и не бывавшие в браке, имеют бОльшее количество задержек выплат (9,3% и 9,8%).
Возможно, узаконенные отношения ведут к более ответственному подходу к своим обязательствам.
Наименьшее количество просроченных платежей среди овдовевших (6,6%), однако в данной категории очень малое количество взятых кредитов (63). Также редко берут кредиты люди в разводе (84). 

3) Влияние уровня дохода на сроки погашения кредита.
Для оценки данного показателя введена классификация по среднемесячному доходу в рублях:  
- 0–30000 — 'E';
- 30001–50000  — 'D';
- 50001–200000 — 'C';
- 200001–1000000 — 'B';
- 1000001 и выше — 'A'.  
Данный показатель незначительно влияет на сроки погашения кредита. Можно отметить, что люди с доходом от 50 до 200 000 руб. оформляют бОльшее количество кредитов. На долю данной категории приходится 78% оформленных кредитов.
Следует учесть, что выборка в категориях с очень низким (до 30 000 руб.) и очень высоким (более 1 млн.руб.)  уровнем дохода крайне мала, что не позволяет сформировать однозначных выводов . 
Рекомендовано  исследовать этот фактор с наличием бОльших данных по категориям А, D и Е. и учитывать влияние уровня дохода в связи с другими факторами.

4) Влияние цели получения на сроки погашения кредита.
По категории  «операции с недвижимостью» оформляется наибольшее количество кредитов и выявлено  наименьшее количество просроченных выплат  (7,3%) . Наименьшее количество кредитов оформлено на проведение свадьбы.
Большее количество просроченных выплат у тех, кому нужен кредит на операции с автомобилем (9,3%) и образование (9,3%).

**Общие рекомендации:** Для построения модели кредитного скоринга возможно использовать данные влияния семейного статуса и цели получения на выполнение кредитных обязательств. Для использования данных о количестве детей и среднемесячном доходе необходимо провести дополнительные исследования по бОльшей выборке данных. Во всех исследованных факторах разница между процентом невыплат в различных категориях не превышает 3%, что не может являться определяющим фактором при принятии решения о кредитоспособности. Также рекомендовано исследование других факторов, присутствующих в исходных данных: возраст, тип занятости.